src/config/enterprise.rs

```diff
@@ -0,0 +1,426 @@
+    host_metrics.scrape_interval_secs(datadog.reporting_interval_secs);
+    internal_metrics.scrape_interval_secs(datadog.reporting_interval_secs);
+
+    config.sources.insert(
+        host_metrics_id.component.clone(),
+        SourceOuter::new(host_metrics),
+    );
+    config.sources.insert(
+        internal_metrics_id.component.clone(),
+        SourceOuter::new(internal_metrics),
+    );
+
+    // Create a Datadog metrics sink to consume and emit internal + host metrics.
+    let datadog_metrics = DatadogMetricsConfig::from_api_key(api_key);
+
+    config.sinks.insert(
+        datadog_metrics_id,
+        SinkOuter::new(
+            vec![host_metrics_id, internal_metrics_id],
+            Box::new(datadog_metrics),
+        ),
+    );
+
+    Ok(())
+}
+
+/// By default, the Datadog feature is enabled.
+const fn default_enabled() -> bool {
+    true
+}
+
+/// By default, Vector should exit when a fatal reporting error is encountered.
+const fn default_exit_on_fatal_error() -> bool {
+    true
+}
+
+/// By default, report to Datadog every 5 seconds.
+const fn default_reporting_interval_secs() -> f64 {
+    5.0
+}
+
+/// By default, keep retrying (recoverable) failed reporting (infinitely, for practical purposes.)
+const fn default_max_retries() -> u32 {
+    u32::MAX
+}
+
+/// By default, retry (recoverable) failed reporting every 5 seconds.
+const fn default_retry_interval_secs() -> u32 {
+    5
+}
+
+/// Returns the full URL endpoint of where to POST a Datadog Vector configuration.
+fn get_reporting_endpoint(site: Option<&String>, configuration_key: &str) -> String {
+    format!(
+        "{}{}/{}/versions",
+        get_api_base_endpoint(None, site, None),
+        DATADOG_REPORTING_PATH_STUB,
+        configuration_key
+    )
+}
+
+/// Build a POST request for reporting a Vector config to Datadog OP.
+fn build_request<'a>(
+    endpoint: &Url,
+    auth: &'a PipelinesAuth,
+    payload: &'a PipelinesVersionPayload,
+) -> Request<Body> {
+    Request::post(endpoint.to_string())
+        .header("DD-API-KEY", auth.api_key)
+        .header("DD-APPLICATION-KEY", auth.application_key)
+        .body(Body::from(payload.json_string()))
+        .unwrap_or_else(|_| {
+            panic!(
+                "couldn't create {} HTTP request. Please report",
+                DATADOG_REPORTING_PRODUCT
+            )
+        })
+}
+
+/// Reports a JSON serialized Vector config to Datadog, for use with Observability Pipelines.
+async fn report_serialized_config_to_datadog<'a>(
+    client: &'a HttpClient,
+    endpoint: &'a str,
+    auth: &'a PipelinesAuth<'a>,
+    payload: &'a PipelinesVersionPayload<'a>,
+) -> Result<(), ReportingError> {
+    info!(
+        "Attempting to report configuration to {}.",
+        DATADOG_REPORTING_PRODUCT
+    );
+
+    let mut endpoint = Url::parse(endpoint).map_err(ReportingError::EndpointError)?;
+
+    // Follow redirection responses a maximum of one time.
+    for _ in 0..=1 {
```

Comment:
Is there a specific reason we limit to 1 redirect? I assume it's since, practically speaking, we won't be redirecting more than once versus just cutting to whatever new `endpoint` format?

Reply:
Exactly. This is a basic security measure to avoid infinite loops. The idea being, since we have control over the DD API, we should be able to ensure a maximum of one hop if the endpoint changes from, say, `/api/unstable` to `/api/v2`.

Now that we're discussing a switch to Rust and GraphQL, this is possibly redundant longer-term, but I think fine to leave in for now.